File: lib/validation/task-input.ts
import type { TaskFormInput, TaskStatus } from "@/lib/domain/types";

const statuses: TaskStatus[] = ["pending", "in_progress", "done"];

export function parseTaskInput(input: unknown): TaskFormInput {
  if (!input || typeof input !== "object") {
    throw new Error("Dados invalidos.");
  }

  const data = input as Partial<Record<keyof TaskFormInput, unknown>>;
  const title = typeof data.title === "string" ? data.title.trim() : "";
  const description = typeof data.description === "string" ? data.description : "";
  const categoryId = typeof data.categoryId === "string" ? data.categoryId : "";
  const status = statuses.includes(data.status as TaskStatus) ? (data.status as TaskStatus) : "pending";
  const dueDate = typeof data.dueDate === "string" ? data.dueDate : "";

  if (!title) {
    throw new Error("Titulo e obrigatorio.");
  }

  if (!categoryId) {
    throw new Error("Categoria e obrigatoria.");
  }

  return {
    title,
    description,
    categoryId,
    status,
    dueDate,
  };
}

File: backend/src/main/java/br/edu/intellitasks/dto/TaskInput.java
package br.edu.intellitasks.dto;

import br.edu.intellitasks.model.TaskStatus;
import java.util.Map;

public record TaskInput(String title, String description, String categoryId, TaskStatus status, String dueDate) {
  public static TaskInput from(Map<String, Object> data) {
    String title = stringValue(data, "title").trim();
    String description = stringValue(data, "description");
    String categoryId = stringValue(data, "categoryId");
    TaskStatus status = parseStatus(stringValue(data, "status"));
    String dueDate = stringValue(data, "dueDate");

    if (title.isBlank()) {
      throw new IllegalArgumentException("Titulo e obrigatorio.");
    }

    if (categoryId.isBlank()) {
      throw new IllegalArgumentException("Categoria e obrigatoria.");
    }

    return new TaskInput(title, description, categoryId, status, dueDate);
  }

  private static TaskStatus parseStatus(String value) {
    try {
      return value.isBlank() ? TaskStatus.pending : TaskStatus.valueOf(value);
    } catch (IllegalArgumentException error) {
      return TaskStatus.pending;
    }
  }

  private static String stringValue(Map<String, Object> data, String key) {
    Object value = data.get(key);
    return value instanceof String ? (String) value : "";
  }
}
